src/dvc_data/hashfile/state.py

```diff
@@ -87,24 +101,42 @@ def close(self):
         self.hashes.close()
         self.links.close()
 
+    @staticmethod
+    def _serialize(path, fs, hash_info):
+        entry = {
+            "checksum": fs.checksum(path),
+            "size": fs.size(path),
+            "hash_info": hash_info.to_dict(),
+        }
+        return json.dumps(entry)
+
     def save(self, path, fs, hash_info):
         """Save hash for the specified path info.
 
         Args:
             path (str): path to save hash for.
             hash_info (HashInfo): hash to save.
         """
+        if not isinstance(fs, LocalFileSystem):
+            return
 
+        self.hashes[path] = self._serialize(path, fs, hash_info)
+
+    @contextmanager
+    def bulk_save(self, fs) -> Iterator[Callable[[str, "HashInfo"], None]]:
```

Comment:
Why bulk_save is a contextmanager that returns a callback rather than one taking iterator/iterables?

It's because if we checkout all of the files first and then try to save state, the values that we save in the state may be newer than the one we checked out. So, we use callback-like mechanism to serialize first and then save that later.